[V1] Enable multiple warmup shapes

Currently the V1 integration will fail when multiple warmup sizes are set with a graph compilation failure. The failure looks very similar to the failure that occurs if you accidentally enable `torch.inference_mode` somewhere. An easy test to trigger this is https://github.com/vllm-project/vllm-spyre/blob/main/tests/test_spyre_warmup_shapes.py

Maybe related to https://github.com/vllm-project/vllm-spyre/issues/12

[This comment](https://github.com/vllm-project/vllm-spyre/pull/11#issuecomment-2711244246) noted that running multiple warmup shapes resluted in a hang; however, when I run `test_spyre_warmup_shapes.py` with v1 (or even with v0), I get this error when using the `sendnn_decoder`:
```
>           raise BackendCompilerFailed(self.compiler_fn, e) from e
E           torch._dynamo.exc.BackendCompilerFailed: backend='sendnn_decoder' raised:
E           AttributeError: 'Node' object has no attribute 'dtype'
E           
E           Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
E           
E           
E           You can suppress this exception and fall back to eager by setting:
E               import torch._dynamo
E               torch._dynamo.config.suppress_errors = True
```
Continuing to investigate...

--- a/tests/e2e/test_spyre_warmup_shapes.py
+++ b/tests/e2e/test_spyre_warmup_shapes.py
@@ -26,7 +26,7 @@ from vllm import SamplingParams
     "warmup_shapes", [[(64, 20, 8),
                        (128, 20, 4)]])  # (prompt_length/new_tokens/batch_size)
 @pytest.mark.parametrize("backend", get_spyre_backend_list())
-@pytest.mark.parametrize("vllm_version", ["V0", "V1"])
+@pytest.mark.parametrize("vllm_version", ["V1"])
 def test_output(
     model: str,
     prompts: list[str],

--- a/vllm_spyre/v1/worker/spyre_model_runner.py
+++ b/vllm_spyre/v1/worker/spyre_model_runner.py
@@ -105,6 +105,9 @@ class SpyreModelRunner(ModelRunnerBase[ModelInputForSpyre]):
         # Lazy initialization: after load_model.
         self.model: nn.Module
 
+        # Flag to be turned off after warmup is complete
+        self.warmup_mode = True
+
     def get_model(self) -> nn.Module:
         return self.model
 
@@ -210,6 +213,10 @@ class SpyreModelRunner(ModelRunnerBase[ModelInputForSpyre]):
                                       dtype=torch.float16,
                                       use_mla=False)
         return {"foo": attn_spec}
+
+    def complete_warmup(self):
+        """Turn off warmup mode once the warmup is complete"""
+        self.warmup_mode = False
 
 
 class StaticBatchingSpyreModelRunner(SpyreModelRunner):
@@ -539,6 +546,10 @@ class StaticBatchingSpyreModelRunner(SpyreModelRunner):
         """Yoinked from 
         https://github.com/foundation-model-stack/aiu-fms-testing-utils/pull/13
         """
+        if not self.warmup_mode:
+            # Only mark tensors when we're warming up and compiling the graphs
+            return
+
         # To produce like graphs during pre-fill, we mark the prefill
         # batch x seq as static, but relax this for decode for the seq
         if model_input.is_prompt:

--- a/vllm_spyre/v1/worker/spyre_model_runner.py
+++ b/vllm_spyre/v1/worker/spyre_model_runner.py
@@ -423,6 +423,7 @@ class StaticBatchingSpyreModelRunner(SpyreModelRunner):
         self._update_states(scheduler_output)
 
         model_input = self.prepare_model_input(scheduler_output)
+        self._mark_input_tensors(model_input)
 
         # TODO(Wallas): I think it would be better move the indices as argument
         # of the forward rather than set as an attribute of the model. I'm not
@@ -533,6 +534,32 @@ class StaticBatchingSpyreModelRunner(SpyreModelRunner):
             'prompt_length']
         padded_batch_size = applicable_spyre_warmup_shapes[0]['batch_size']
         return padded_batch_size, min_pad_length_batch
+
+    def _mark_input_tensors(self, model_input: ModelInputForSpyre) -> None:
+        """Yoinked from 
+        https://github.com/foundation-model-stack/aiu-fms-testing-utils/pull/13
+        """
+        # To produce like graphs during pre-fill, we mark the prefill
+        # batch x seq as static, but relax this for decode for the seq
+        if model_input.is_prompt:
+            # we always want prefill to be static to produce same-like graph
+            torch._dynamo.mark_static(model_input.input_tokens, 0)
+            torch._dynamo.mark_static(model_input.input_tokens, 1)
+            torch._dynamo.mark_static(model_input.input_masks, 0)
+            torch._dynamo.mark_static(model_input.input_masks, 1)
+            torch._dynamo.mark_static(model_input.input_masks, 2)
+            torch._dynamo.mark_static(model_input.input_positions, 0)
+            torch._dynamo.mark_static(model_input.input_positions, 1)
+        else:
+            # we always want the decode to be dynamic on sequence
+            torch._dynamo.mark_dynamic(model_input.input_tokens, 1)
+            torch._dynamo.mark_dynamic(model_input.input_masks, 1)
+            torch._dynamo.mark_dynamic(model_input.input_masks, 2)
+
+            # here self.model.model is a StaticBatchingFmsModel
+            for layer in self.model.model.past_key_value_states:
+                for tensor in layer:
+                    torch._dynamo.mark_static(tensor, 0)
 
 
 class ContinuousBatchingSpyreModelRunner(SpyreModelRunner):

--- a/vllm_spyre/v1/worker/spyre_worker.py
+++ b/vllm_spyre/v1/worker/spyre_worker.py
@@ -88,6 +88,7 @@ class SpyreWorker(WorkerBaseV1):
             "All warmups for %d different prompt/decode/batchsize-shape "
             "combinations finished. Total warmup time %.3fs.",
             len(wup_new_tokens), all_warmup_total_t)
+        self.model_runner.complete_warmup()
 
     def check_health(self) -> None:
         """Basic health check (override for device-specific checks)."""

